docs/consensus_decisions.md

```diff
@@ -0,0 +1,51 @@
+
+One of the drawbacks of message references is that referenced messages effectively have a public API (names of parameters, variables, variants, etc.) which must be consistent across all callsites.
+This leads us to consensus 2.
+
+## 2: Allow parameters passed with message references to the message being referenced and validate it.
+
+**Discussion:**
+The variables/fields passed should not be completely untyped and unchecked.
+We want a validation mechanism that can allow providing early error feedback to the translators & developers.
+We need to decide on when the validation can & should happen, including the meaning of “build time” and “run time” in regards to validation.
+
+## 3: Allow for selectors to select a case depending on the value of one or more input arguments.
+
+**Discussion:**
+This is a prerequisite for top-level selectors to be able to represent complex messages, without requiring those messages to be split up in an unergonomic manner.
+This is an extension or relaxation of what's allowed in MessageFromat 1.
+
+While message references make it technically possible for the data model to represent multi-argument selectors otherwise, this requires the use of n²-1 artificial "messages", where n is the number of arguments. This is not desirable.
+
+## 4: Only allow for selectors at the top level of a message.
+
+**Discussion:**
+Requiring selectors to only be available at the top level is a good way of helping to maintain the translatability of messages, as well as otherwise guiding MessageFormat 2 users towards good practices.
+
+After an in-depth exploration of the problem space, we have determined that while selectors are a necessary feature of MessageFormat, it is not necessary for them to be available within the body of a message, or directly within a case of a parent selector.
+
+All identified use cases of such constructions may be cleanly represented using a top-level selector that may use more than one input argument to select among a set of messages.
+Furthermore, we may enable complete reversibility of message transformations to and from languages such as MessageFormat 1 and Fluent by using message references.
+
+## 5: Top level selectors together with message references provides the same value as nested selectors at a lower cost.
+
+**Discussion:**
+Nested selectors provides capabilities that may be useful in avoiding variant permutation explosion in edge cases, but the use of them has not been evaluated in production localization systems to date.
+The group believes that the known value of this feature can be sufficiently covered by the combination of message references and top level selection features, which together provide a sufficient feature set at a lower cost to the ecosystem than nested selectors would do.
+
+## 6: The group will attempt to avoid blocking addition of nested selectors in the future revisions of the standard.
```

Comment:
I think something like "The group does not intend to block addition of..." would be a little clearer here.

Reply:
@zbraniecki, would you be comfortable with this change? I'm not entirely sure whether the change in phrasing changes the intent here.

Reply:
I don't think it captures the spirit of my intent as well as the original version, since the original version expresses an intent to avoid an action (`we will try to avoid X`), while the revision expresses lack of intent for an opposite action (`we do not intend to do Y` where `Y = !X`), but at this point, I feel like the amount of energy invested in this point is starting to outweigh the original value of this point :)

If Dan feels this is a better sentence, and Eemeli agrees, I'm ok with that.

Reply:
Sorry, I didn't mean to stir the pot here. If it changes the intent, even slightly, then we should probably go with the original wording.

Reply:
Let's keep the original then.

Reply:
Ping @stasm, @zbraniecki: Just to be clear, this is the change that's now intended, yes?

Reply:
yes.

Reply:
I wasn't involved in this discussion, so I don't know the background, and it sounds quite contentious, but it seems that the group decided not to include X in the first version of the standard (where X = nested selectors) on the condition that they would be considered given real world experience. I think it is going a bit far to declare that a future standards body should "avoid" blocking X, since that phrasing seems to imply that the body in 2021 wants to short-circuit debate in the future. So I prefer the new wording.

I would personally remove the double-negativd from the bullet point and say simply: "The group plans to revaluate nested selectors given real world experience."

Reply:
Like @sffc I wasn't involved in this discussion, and am playing catch-up here. I don't want to derail a meaningful consensus that I might agree with. But...

I'm concerned about the statement in question since at Amazon we use nested plural/select statements often and I have "evaluated it in a production localization system" :-). While we discourage deep nesting (more than about 3 levels) and our tools/formats are harder to use with deep nesting (it hasn't been a priority for us to work on it), we nevertheless currently allow an arbitrary level of select/plural nesting. I have used multiple selects to good effect in cases too numerous for me to ignore. So I agree that we should not attempt to block the addition of nested selectors--in fact, I think I'd insist on the ability to do multiple levels.

The challenges of select/plural nesting when applied to string fragments (vs. forcing "complete thought string" repetition) I've enumerated elsewhere, but I am violently opposed to the original wording of item 6. I freely admit this is a hard problem: we should not shy away from hard problems.

Reply:
@sffc i feel like we did not commit to reevaluate it as a group. My intention for this point was to give a clue to some future group that will be evaluating it that we did, and the fact that we didn't decide to include it was not because we don't see their value.
I also wanted to put a bit of a constrain on ourselves for the remainder of this work group existence to try not to lock the spec from being able to be extended with this feature.

@stasm seems to be concerned that it is requesting us to "sign up for something" that he doesn't understand. I wish I had energy to solve it by improving understanding of what I'd like us to commit to, but j frankly don't have it.

@aphillips I'm your ally here. I really am. For months. And unfortunately the strongest thing i was able to almost get a consensus on was the original wording.
If you are violently opposing it, may I ask what wording would you suggest?

Reply:
May I also point out that it's really sad that a major stakeholder org like Amazon has not been represented in the task force group evaluating nested selectors.
I think something like this should be explicitly stated as a malfunction of our work group.

Reply:
@zbraniecki @dminor @eemeli  Thanks, that helps clarify. How about this then: "The group is not ruling out consideration of nested selectors in the future."

(I don't have anything to say to @aphillips; I'm only iterating on the phrasing discussion)

Reply:
@zbraniecki it's more my/Amazon's fault for not participating actively enough or pushing on the WG when the timing of meetings became a problem than the WGs fault.

I think I may have misread the original text, but I tend to favor @sffc's suggestion "The group is not ruling out consideration of nested selectors in the future."

Reply:
> I think I may have misread the original text, but I tend to favor @sffc's suggestion "The group is not ruling out consideration of nested selectors in the future."

I see four possible messages we may want to convey:

1) This group would like the reader from the future to know, that it's decision to not include nested selectors was driven by insufficient justification at the time and should not be interpreted as a rejection of the concept.
2) The group may revisit nested selectors in the future.
3) The group conducted a lot of research and brainstorming on nested selectors and the group saw value in the future, but not sufficient to invest in it at the time.
4) The group would like the reader to know that the group would like the reader, if the reader is involved in the work on MF2.0, to put effort in ensuring that the standard will not end up making extension of nested selectors impossible.

My original statement was primarily focused on (4), with some of (3) and (1).
The @dminor's revision switched it to primarily (1) with a bit of implicit (2) and (4).
@sffc's version switches to primarily (2).

I think my aim to achieve (4) is futile, and I do not wish to invest this group's time in trying to recognize the value of my goal.

Reply:
I have gone back and forth in my mind for whether internal selectors should or should not be allowed. I found the current wording acceptable for now.

On one side, traditional translators tend to not simplify logic and prefer duplicating a translation. They seem to appreciate seeing the entire sentence in one line.  My anecdotal evidence is that it lowers their cognitive load during translation. It's also easier to detect message changes because when the selector is in the middle, it may be harder to diff the changes when updating a translation of a message.

On the other side, message authors will need to modify more complicated messages. If they need to modify a word or two, the inability to have internal selectors can be rather burdensome because the same word will need to be modified across several selection statements. I know of one message with at least 10 **optional** arguments that we have. Though it's closer to about 20 arguments if you include the non-optional arguments. It's one of the worst messages that our translators ever have to touch. You could say that the author shouldn't have created such a monstrous message. The rebuttal to such an argument is typically that such a message is required, and that the technology should support such things. If it didn't have internal selectors, then they would probably resort to concatenating the string in code to get what they wanted, and that would thus put the translators in a worse position.

So I'm in conflict, and I'm okay with a "we're not saying no" approach for now. If no internal selectors for now allows for us to work on other important things, then I'm okay with that. Allowing us to detect and select grammatical properties of words or conveying context dependencies between selectors is of the utmost importance for my goals.

Reply:
@zbraniecki - my intent was to convey mainly points 1 and 3. The existing explanation below the topic sentence explains this pretty clearly. However, I can see that we want to also convey point 4.

I think maybe my confusion arises from the phrase "avoid blocking".  I interpret that double-negative to mean that the *standards body* will avoid blocking, in the **procedural** sense of the word.  However, it appears that this is the wrong sense of the word "block"; do you intend to mean it in the sense that we should not make *design decisions* that would *prevent* a potential future addition of the feature?

How about this: "The model will be designed the door open for nested selectors to be potentially reconsidered in the future."

Reply:
@grhoten As planned, the technology will support even your "monstrous message" via message references. While I'm not advocating that the following is a good idea, it's at the very least possible based on the consensus decisions included in this PR.

This is a third equivalent expression of the message provided by @aphillips in [another thread on this PR](#discussion_r584185434), this time optimised for lack of repetition of content. The syntax is a representation of one of the data model candidates we're developing.

```
{
  id: 'unread-somethings',
  value: [
    'Hello ',
    { var_path: ['name'] },
    ' you have ',
    { msg_path: ['unread-somethings-count'] },
    ' in your inbox.'
  ]
},

{
  id: 'unread-somethings-count',
  value: {
    select: [{ func: 'plural', args: [{ var_path: ['unreadCount'] }] }],
    cases: [
      {
        key: [0],
        value: [' no ', { msg_path: ['something', { var_path: ['msgType'] }, 'plural'] }]
      },
      {
        key: ['one'],
        value: [
          { var_path: ['unreadCount'] },
          ' ',
          { msg_path: ['something', { var_path: ['msgType'] }, 'single'] }
        ]
      },
      {
        key: ['other'],
        value: [
          { var_path: ['unreadCount'] },
          ' ',
          { msg_path: ['something', { var_path: ['msgType'] }, 'plural'] }
        ]
      }
    ]
  }
},

{
  id: 'something',
  entries: [
    {
      id: 'email',
      entries: [
        { id: 'single', value: ['email'] },
        { id: 'plural', value: ['emails'] }
      ]
    },

    {
      id: 'notification',
      entries: [
        { id: 'single', value: ['notification'] },
        { id: 'plural', value: ['notifications'] }
      ]
    },

    {
      id: 'other',
      entries: [
        { id: 'single', value: ['item'] },
        { id: 'plural', value: ['items'] }
      ]
    }
  ]
}
```

Reply:
> I think maybe my confusion arises from the phrase "avoid blocking". I interpret that double-negative to mean that the standards body will avoid blocking, in the procedural sense of the word. However, it appears that this is the wrong sense of the word "block"; do you intend to mean it in the sense that we should not make design decisions that would prevent a potential future addition of the feature?

Ah, I can see that confusion now! Thank you for pointing it out.
Yes, my intended meaning of the word "blocking" was technical, not political.

> How about this: "The model will be designed to leave the door open for nested selectors being potentially reconsidered in the future."

That sounds great and is even stronger than my original one. I am happy with a stronger form because I really believe that we should keep the door open.

Reply:
```suggestion
## 6: The model will be designed to leave the door open for nested selectors being potentially reconsidered in the future.
```

@stasm, would this form suggested by @sffc and seconded by @zbraniecki be acceptable to you?

Reply:
Thanks for the additional discussion, everyone. I think it was the word _blocking_ which I wasn't sure about. The new wording is much clearer, thanks!

Reply:
And yes, I also agree that the new wording is stronger, and I'm OK with it. My main issue with the previous wording was that I wasn't really sure what it entailed. The expectations of the new wording are easier for me to understand.

Reply:
This isn't resolved yet.  It looks like dminor's wording got committed, instead of the new wording I proposed above.

Reply:
Oh bother, I clicked the wrong button and messed up. Apologies. Fixing now, and then re-marking this thread as resolved.